### PR TITLE
Bug 1811834: Sync jsonnet dependencies

### PR DESCRIPTION
**Pull request by paulfantom**
Multiple things in one PR (sorry):
- Porting https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/375 to fix [BZ 1811834](https://bugzilla.redhat.com/show_bug.cgi?id=1811834)
- Securing `/metrics` endpoint for prometheus-operators, [JIRA MON-990](https://issues.redhat.com/browse/MON-990)
- updating labels on node-exporter and adding mechanism to delete DaemonSet when necessary

Some other changes include ones described in https://github.com/coreos/kube-prometheus/pull/470

Note: `jsonnetfile.lock.json` was updated in stages to check what is changing, but in the end it came down to running `jb update` on all.

/cc @openshift/openshift-team-monitoring

### CODE REVIEW

**Comment by paulfantom** [comment]
/hold
until https://github.com/prometheus/node_exporter/pull/1649 is merged and propagated

**Comment by lilic** [comment]
Its been merged it seems.

**Comment by paulfantom** [comment]
/unhold

https://github.com/prometheus/node_exporter/pull/1649 is merged and propagated here.

File: assets/prometheus-operator/service-monitor.yaml [hunk 1]
```diff
@@ -9,8 +9,19 @@ metadata:
   namespace: openshift-monitoring
 spec:
   endpoints:
-  - honorLabels: true
-    port: http
+  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
```
Comment (simonpasquier) [inline assets/prometheus-operator/service-monitor.yaml]: why do we have 2 endpoint entries?

Comment (paulfantom) [reply assets/prometheus-operator/service-monitor.yaml]: Good catch! Fixing.

File: pkg/client/client.go [hunk 2]
```diff
@@ -783,7 +783,22 @@ func (c *Client) CreateOrUpdateDaemonSet(ds *appsv1.DaemonSet) error {
 	}
 
 	err = c.UpdateDaemonSet(ds)
-	return errors.Wrap(err, "updating DaemonSet object failed")
+	if err != nil {
+		uErr, ok := err.(*apierrors.StatusError)
+		if ok && uErr.ErrStatus.Code == 422 && uErr.ErrStatus.Reason == metav1.StatusReasonInvalid {
+			// try to delete DaemonSet
+			err = c.CreateDaemonSet(ds)
```
Comment (simonpasquier) [inline pkg/client/client.go]: `c.DeleteDaemonSet(ds)` rather?

Comment (paulfantom) [reply pkg/client/client.go]: fixed

**Comment by lilic** [comment]
/retest

should be back now

**Comment by paulfantom** [comment]
Rebased to solve merge conflicts

File: assets/prometheus-k8s/rules.yaml [hunk 3]
```diff
@@ -804,6 +804,44 @@ spec:
...
+      for: 10m
+      labels:
+        severity: warning
+    - alert: NodeClockNotSynchronising
+      annotations:
+        message: Clock on {{ $labels.instance }} is not synchronising. Ensure NTP
+          is configured on this host.
+        summary: Clock not synchronising.
+      expr: |
+        min_over_time(node_timex_sync_status[5m]) == 0
```
Comment (lilic) [inline assets/prometheus-k8s/rules.yaml]: We don't need the job label selector here, like we had for ClockSkewDetected?

File: assets/prometheus-operator-user-workload/cluster-role.yaml [hunk 4]
```diff
@@ -87,3 +87,15 @@ rules:
   - get
   - list
   - watch
+- apiGroups:
```
Comment (lilic) [inline assets/prometheus-operator-user-workload/cluster-role.yaml]: Do we know why this was added for prometheus-operator in kube-prom, can't seem to find it in the prometheus-operator repo itself? https://github.com/coreos/prometheus-operator/blob/master/example/rbac/prometheus-operator/prometheus-operator-cluster-role.yaml should this be updated there?

File: pkg/client/client.go [hunk 5]
```diff
@@ -632,17 +631,17 @@ func (c *Client) CreateOrUpdateDeployment(dep *appsv1.Deployment) error {
...
+				return errors.Wrap(err, "deleting Deployment object failed")
 			}
 			err = c.CreateDeployment(dep)
 			if err != nil {
-				return errors.Wrap(err, "creating deployment object failed after update failed")
+				return errors.Wrap(err, "creating Deployment object failed after update failed")
 			}
 		}
-		return errors.Wrap(err, "updating deployment object failed")
+		return errors.Wrap(err, "updating Deployment object failed")
```
Comment (lilic) [inline pkg/client/client.go]: Thanks :)

Comment (lilic) [reply pkg/client/client.go]: Follow up task to us, maybe tech-debt, unrelated to this PR: we should include the name of these objects in the logs. WDYT?

File: assets/prometheus-k8s/rules.yaml (hunk 3 above)
Comment (paulfantom) [reply assets/prometheus-k8s/rules.yaml]: That metric comes only from node_exporter which is deployed as DaemonSet, so I don't see why we would need it.

File: assets/prometheus-operator-user-workload/cluster-role.yaml (hunk 4 above)
Comment (paulfantom) [reply assets/prometheus-operator-user-workload/cluster-role.yaml]: Without those kube-rbac-proxy didn't work and metrics could be scraped.

File: pkg/client/client.go (hunk 5 above)
Comment (paulfantom) [reply pkg/client/client.go]: good idea, could you create JIRA?

File: assets/prometheus-operator-user-workload/cluster-role.yaml (hunk 4 above)
Comment (lilic) [reply assets/prometheus-operator-user-workload/cluster-role.yaml]: okay that explains that, thanks!

**Comment by lilic** [comment]
aws fails with 🤔:
> fail [github.com/openshift/origin/test/extended/prometheus/prometheus.go:231]: possibly some services didn't register ServiceMonitors to allow metrics collection
Unexpected error:
    <*errors.errorString | 0xc00028d1d0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred

**Comment by paulfantom** [comment]
argh, yes. We are checking if prometheus-operator is available over http and not https -.-

Source: https://github.com/openshift/origin/blob/master/test/extended/prometheus/prometheus.go#L222

File: pkg/client/client.go (hunk 5 above)
Comment (lilic) [reply pkg/client/client.go]: Done! https://issues.redhat.com/browse/MON-1048

**Comment by lilic** [comment]
@sferich888 that won't work as the lock file needs to be generated for each release branch, what are you trying to do backport? :)

**Pull request merged**